airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -400,37 +407,47 @@ def _get_checkpoint_reader(
         sync_mode: SyncMode,
         stream_state: MutableMapping[str, Any],
     ) -> CheckpointReader:
-        checkpoint_mode = self._checkpoint_mode
+        mappings_or_slices = self.stream_slices(
+            cursor_field=cursor_field,
+            sync_mode=sync_mode,  # todo: change this interface to no longer rely on sync_mode for behavior
+            stream_state=stream_state,
+        )
+
+        # Because of poor foresight, we wrote the default Stream.stream_slices() method to return [None] which is confusing and
+        # has now normalized this behavior for connector developers. Now some connectors return [None]. This is objectively
+        # misleading and a more ideal interface is [{}] to indicate we still want to iterate over one slice, but with no
+        # specific slice values. None is bad, and now I feel bad that I have to write this hack.
+        # if mappings_or_slices == [None] or mappings_or_slices == []:
+        if mappings_or_slices == [None]:
+            mappings_or_slices = iter([{}])
+
+        slices_iterable_copy, iterable_for_detecting_format = itertools.tee(mappings_or_slices, 2)
+        legacy_slice_format, is_substream = self._detect_stream_category(mappings_or_slices=iterable_for_detecting_format)
+        self.is_substream = is_substream
+
+        self.cursor = None  # todo: This doesn't work for file-based streams because it doesn't support cursor reassignment. Need to fix
```

Comment:
why do we need to set `self.cursor` to None?

Reply:
This was because we initially would end up setting the cursor in the availability strategy and so when we do the actual read we would end up with the wrong cursor. However, I found a _slightly_ better way of assigning cursor more intuitively which I'll push in my next commit and explain in a comment in the code